### DISCUSSION TITLE
How to configure the client certificate key?

### DISCUSSION THREAD

**Discussion by Tk8and15** [geldata/gel#4739] [Q&A]
Hello my friend, do you have a TypeScript process for connecting to clients? I can't understand his document, can you share it?
In which file is it configured? How to connect the client with TypeScript? The document is too concise, I don't understand it, can you help me?

**Comment by raddevon** [comment]
If you're connecting the TS client within a directory where you've run `edgedb project init`, the client should automatically connect to the correct instance and database. If you're trying to connect to a different instance, you can either set the `EDGEDB_DSN` environment variable to a valid DSN for your instance or pass the DSN when you create the client.

If you need to provide a certificate for connection, you can store that in a JSON file (detailed in [our reference on connecting to an instance](https://www.edgedb.com/docs/reference/connection#specifying-an-instance)) and point to that file when you [create your client](https://www.edgedb.com/docs/clients/js/reference#createClient). You'll need to pass an options object with a `credentialsFile` key containing the path to your credentials file.

**Reply by Tk8and15** [reply]
Where can I get this certificate? Where is the JSON file, do you need to start it?

**Comment by Tk8and15** [comment]
Are there two ways to configure DNS? One is to set the environment variable to a valid DNS, and the other is to pass the DSN when creating the client. How to set up the first one, can you be more specific? Can you give me an example? Then the latter is the error that occurred when I created the client, can you please help me see? In addition, I don't know how to get its certificates and keys, and I don't know how to configure it.
![image](https://user-images.githubusercontent.com/118907993/204940608-87954629-dae3-49ad-a0e3-08b83796217e.png)

![image](https://user-images.githubusercontent.com/118907993/204939782-8a1307e0-59a5-4c08-b0e1-d68b2e9e675e.png)

**Reply by raddevon** [reply]
Yes, there are two ways to configure the DSN, but either one should work. You don't have to use both.

How did you create this instance? Was it created via the CLI with `edgedb project init`?

**Reply by Tk8and15** [reply]
Yes, I am using CLI edgedb project init. Can you tell me how to configure DSN and self-signed certificate?

**Reply by raddevon** [reply]
If this script is also inside that project you created via `edgedb project init`, the easiest way to get it to connect would be to stop passing that options object to the `createClient` function. EdgeDB will handle connecting to the correct database and supplying the certificate without your having to configure it manually while you're within a project.

So, your `createClient` call would look like this:

```
const baseClient = createClient();
```

**Reply by Tk8and15** [reply]
Thank you very much for telling me that I used your method at first, but it will report an error and tell me that a DSN needs to be configured. I don't know how to solve it. I also know that an unknown certificate will be generated when I use the edgedb project init, but I don't know what use that certificate will have?

**Reply by Tk8and15** [reply]
At the beginning of my project, I used its quick entry case, which contained Movie, but I couldn't find Movie when I used the query quotation in the client. Will prompt me to say that my self-visa certificate is invalid.

**Reply by raddevon** [reply]
It seems like you may have just gotten your project into a bad state somehow. Let's try to reset and start from scratch.

Try this:

1. Run `edgedb project unlink -D` inside your project to destroy the project.
2. Run `edgedb project init` again to re-initialize a new project with a new instance.
3. Try again with the bare `createClient` call without any options. That should work on a freshly created project/instance as long as the script is below where you ran `edgedb project init` in your directory tree.

**Reply by Tk8and15** [reply]
OK, thank you for answering my questions. I will try your method. If it doesn't work, I'll ask you again. I'm sorry to bother you.

**Reply by raddevon** [reply]
You're welcome. It's not a bother. That's what I'm here for. 😊

**Reply by Tk8and15** [reply]
I hope I can master edgedb with your help as soon as possible

**Reply by Tk8and15** [reply]
This is my edgedb project init error or the same problem, here is my configuration and documentation.
![MGKOOS WC~P~ ~2IY P0ZT7](https://user-images.githubusercontent.com/118907993/205208638-3b3a65e1-3e56-4acb-b6fe-ffabe5f2e977.png)
![@7NW%O47Z%U)SJZ5T(E{%}X](https://user-images.githubusercontent.com/118907993/205208828-c5a6b047-b560-4a22-bb25-4f3233730fbb.png)
![image](https://user-images.githubusercontent.com/118907993/205208840-00ceed27-284b-4ca8-bb49-46dd126bd6d5.png)
![image](https://user-images.githubusercontent.com/118907993/205208917-3be46abd-1acf-45ce-9f18-1bcc0064d4ad.png)
![image](https://user-images.githubusercontent.com/118907993/205208918-a9b70674-f42b-421a-86fe-1bd9150f889b.png)

**Reply by tailhook** [reply]
Does running a REPL in the command-line tool works? I.e. running just `edgedb` from the project directory?

**Reply by Tk8and15** [reply]
Hello sir, glad we meet again, my remote instance client is linked, but my local client is not.

**Reply by raddevon** [reply]
@Tk8and15 Try running `edgedb configure set listen_addresses 127.0.0.1 ::1` and then start your app. Looks like the app is trying to connect on `::1`. EdgeDB may not be listening on that address.

**Reply by Tk8and15** [reply]
Today, the remote project client can connect, but the local one can't. The remote client can insert data and query data, but the local client connection still reports the same error.

**Reply by raddevon** [reply]
@Tk8and15 OK. So, you tried the command I suggested (`edgedb configure set listen_addresses 127.0.0.1 ::1`), but you're still getting the same error?

Did you see @tailhook's question about whether you can connect locally using just `edgedb`? Can you get a console into the local database using that command?

**Reply by Tk8and15** [reply]
I saw it, and I tried it, but it's still the same.

**Reply by raddevon** [reply]
You saw what, and you tried what? Did you try connecting via `edgedb`? Did you try configuring the listen address? Did you try both?

**Reply by Tk8and15** [reply]
Here is my configuration:
![8TZY8XSA$I7MW@6`Y{GFVSS](https://user-images.githubusercontent.com/118907993/205417215-05e35e81-bb44-4f08-bd17-ed84092f3a36.png)
![image](https://user-images.githubusercontent.com/118907993/205417262-d296e04d-430d-4195-adf0-e75c5310e1db.png)
The remote display is out. This unit is not displayed:
![image](https://user-images.githubusercontent.com/118907993/205417345-004fc157-e9bf-46cc-806f-d25b2a6f66e4.png)
Running the native port is still the same error:
![image](https://user-images.githubusercontent.com/118907993/205417398-c71a787e-be09-4845-89b3-57b0ca41c6dc.png)

**Reply by raddevon** [reply]
What's the output if you run `edgedb` from the project root?
